README.rst

| |travisci| |version| |downloads| |versions| |impls| |wheel| |coverage| |br-coverage|

.. |travisci| image:: https://api.travis-ci.org/jonathaneunice/intspan.svg
    :target: http://travis-ci.org/jonathaneunice/intspan

.. |version| image:: http://img.shields.io/pypi/v/intspan.svg?style=flat
    :alt: PyPI Package latest release
    :target: https://pypi.python.org/pypi/intspan

.. |downloads| image:: http://img.shields.io/pypi/dm/intspan.svg?style=flat
    :alt: PyPI Package monthly downloads
    :target: https://pypi.python.org/pypi/intspan

.. |versions| image:: https://img.shields.io/pypi/pyversions/intspan.svg
    :alt: Supported versions
    :target: https://pypi.python.org/pypi/intspan

.. |impls| image:: https://img.shields.io/pypi/implementation/intspan.svg
    :alt: Supported implementations
    :target: https://pypi.python.org/pypi/intspan

.. |wheel| image:: https://img.shields.io/pypi/wheel/intspan.svg
    :alt: Wheel packaging support
    :target: https://pypi.python.org/pypi/intspan

.. |coverage| image:: https://img.shields.io/badge/test_coverage-100%25-6600CC.svg
    :alt: Test line coverage
    :target: https://pypi.python.org/pypi/intspan

.. |br-coverage| image:: https://img.shields.io/badge/branch_coverage-100%25-6600CC.svg
    :alt: Test branch coverage
    :target: https://pypi.python.org/pypi/intspan

``intspan`` is a ``set`` subclass that conveniently stores sets of integers.
Sets can be created from and displayed as integer spans such as
``1-3,14,29,92-97`` rather than exhaustive member listings. Compare::

    intspan('1-3,14,29,92-97')
    [1, 2, 3, 14, 29, 92, 93, 94, 95, 96, 97]

While they indicate the same values, the ``intspan`` is more compact.
Even more important, it
better divulges the contiguous nature of parts of the collection. It
is easier for humans to quickly determine the "shape" of the collection
and ascertain "what's missing?"

When iterating, ``pop()``-ing an item, or converting to a list, ``intspan``
behaves as if it were an ordered--in fact, sorted--collection. A key
implication is that, regardless of the order in which items are added,
an ``intspan`` will always be rendered in the most compact, organized
form possible.

The main draw is having a convenient way to specify (possibly discontinuous)
ranges--for example, rows to process in a spreadsheet. It can also help you
quickly identify or report which items were *not* successfully processed in
a large dataset.

There is also an ordered ``intspanlist`` type that helps specify the
ordering of a set of elements.

For this and more, see the full details on `Read the Docs
<http://intspan.readthedocs.org/en/latest/>`_.


CHANGES.rst


Change Log
==========

 *  Version 1.4.1 achieves 100% *branch* coverage. *Hooah!*

 *  Version 1.4.0 achieves 100% test coverage and integrated coverage testing
    across multiple versions with ``tox``.

 *  Version 1.3.10 improved test coverage and explicit coverage testing.
    Also tweaked docs.

 *  Version 1.3.9 simplified ``setup.py`` and packaging.

 *  Version 1.3.7 adds ``bdist_wheel`` packaging support.

 *  Version 1.3.6 switches from BSD to Apache License 2.0 and integrates
    ``tox`` testing with ``setup.py``

 *  Version 1.3 adds ``*`` notation for abstract "the rest of the items"
    in an ``intspanlist``.

 *  Version 1.2.6 inaugurates continuous integration with Travis CI.

 *  Version 1.2 adds an experimental ``spanlist`` constructor and
    ``intspanlist`` type.

 *  Version 1.1.0 adds ``from_range`` and ``complement`` methods; improves
    error handling of ``pop`` on an empty set), and tweaks testing.

 *  Patch versions through 1.0.3 are minor bumps, with small testing and
    documentation improvements.

 *  Version 1.0 immediately follows 0.73. Bumped to institute a
    cleaner "semantic versioning" scheme. Upgraded from "beta" to
    "production" status.

 *  Version 0.73 updates testing to include the latest Python 3.4

 *  Version 0.7 fixed parsing of spans including negative numbers, and
    added the ``ranges()`` method. As of 0.71, the ``from_ranges()``
    constructor appeared.


test/test_intspan.py

import pytest
from intspan import *
from intspan import ParseError


def test_basic():
    s = intspan()
    tests = ['', '1', '1-2', '1-3,9-10', '1-3,14,29,92-97']
    for t in tests:
        s = intspan(t)
        assert str(s) == t


def test_alt_contstructors():
    assert intspan(range(100)) == intspan('0-99')
    assert intspan([1,3,5]) == intspan('1,3,5')
    assert intspan([5,3,1]) == intspan('1,3,5')
    assert intspan(intspan('1,3,5')) == intspan('1,3,5')


def test_parse_error():
    with pytest.raises(ParseError):
        s = intspan('7*99')
    with pytest.raises(ParseError):
        s = intspan('1-4,5-')


def test_spaces():
    assert list(intspan('  ')) == []
    assert list(intspan('     1')) == [1]
    assert list(intspan('1, 4,  4 , 9')) == [1, 4, 9]
    assert list(intspan('1, 4, 4 , 9')) == [1, 4, 9]
    assert list(intspan('1, 4, 4 , 9')) == [1, 4, 9]
    assert list(intspan('   1, \n4,\n 4 , 9')) == [1, 4, 9]


def test_negatives():
    assert list(intspan('-2')) == [-2]
    assert list(intspan('-2-1')) == [-2, -1, 0, 1]
    assert list(intspan('-2--1')) == [-2, -1]


def test_contains():
    s = intspan()
    assert 1 not in s
    assert 100 not in s
    assert 0 not in s

    t = intspan('1,10')
    assert 1 in t
    assert 10 in t
    assert 0 not in t
    assert 2 not in t


def test_equals():
    s = intspan('1,3,5,7,9')
    assert s == set([1, 3, 5, 7, 9])


def test_strict_super_or_subset():
    s = intspan('1,3,5,7,9')
    t = intspan('1,3,5')
    u = intspan('0,1,3,5')
    assert s > t
    assert not s > u
    assert t < s
    assert t < u
    assert u > t
    assert not s < u
    assert not s > u


def test_isdisjoint():
    s = intspan('1,3,5,7,9')
    t = intspan('33-44')
    u = intspan('1,3,99,299')
    assert s.isdisjoint(t)
    assert not s.isdisjoint(u)
    assert t.isdisjoint(u)


def test_copy():
    t = intspan('1,10')
    tt = t.copy()
    assert type(tt) == type(t)
    assert t == tt
    assert t is not tt


def test_clear():
    s = intspan('1,2,3,5,8,13,21')
    s.clear()
    assert s == intspan()


def test_len():
    s = intspan('1,2,3,5,8,13,21')
    assert len(s) == 7
    s.pop()
    assert len(s) == 6
    s.clear()
    assert len(s) == 0


def test_merge():
    assert str(intspan('1-4,5')) == '1-5'


def test_out_of_order():
    assert str(intspan('1,0,99,4,7,9,98')) == '0-1,4,7,9,98-99'


def test_discard():
    s = intspan('1-3,14,29,92-97')
    s.discard('2,13,92')
    assert str(s) == '1,3,14,29,93-97'


def test_remove():
    s = intspan('1-3,14,29,92-97')
    s.remove('2,92')
    assert str(s) == '1,3,14,29,93-97'
    with pytest.raises(KeyError):
        s.remove(1000)


def test_add():
    s = intspan('1-2')
    s.add('3,29')
    assert str(s) == '1-3,29'
    s.add('92,97,96,95,94')
    assert str(s) == '1-3,29,92,94-97'
    s.add(93)
    assert str(s) == '1-3,29,92-97'
    s.add('14')
    assert str(s) == '1-3,14,29,92-97'


def test_iteration():
    s = intspan('92,97,96,95,0,94')
    assert [item for item in s] == [0, 92, 94, 95, 96, 97]
    assert list(s) == [0, 92, 94, 95, 96, 97]
    assert set(s) == set([0, 92, 94, 95, 96, 97])


def test_issubset():
    s = intspan('92,97,96,95,0,94')
    assert s.issubset('0-100')
    assert s.issubset(range(98))
    assert s.issubset(range(101))
    assert s.issubset('0, 92-100')
    assert s.issubset([0] + list(range(92, 101)))
    assert s.issubset(intspan('92,97,96,95,0,94'))
    assert s.issubset([0, 92, 94, 95, 96, 97])
    assert not s.issubset('0-10')
    assert not s.issubset(range(20))
    assert not s.issubset(range(95))


def test_issuperset():
    s = intspan('0-3,7')
    assert s.issuperset('0-2')
    assert s.issuperset([0, 1, 3])
    assert not s.issuperset(range(6))
    assert not s.issuperset('0-6')

    assert s >= intspan('0-2')
    assert s >= intspan([0, 1, 3])
    assert not s >= range(6)
    assert not s >= intspan('0-6')


def test_union():
    s = intspan('0-3,7')
    assert s.union('0-2') == s
    assert list(s.union('0-2')) == [0, 1, 2, 3, 7]
    assert list(s.union([99, 101])) == [0, 1, 2, 3, 7, 99, 101]
    assert s.union([99, 101]) == intspan('0-3,7,99,101')

    assert s | intspan('0-2') == s.union('0-2')
    assert s | [99, 101] == s.union('99,101')


def test_intersection():
    s = intspan('1-8')
    t = intspan('2-5')
    u = intspan('8,100')
    assert s.intersection(t) == intspan('2-5')
    assert t.intersection(u) == intspan()
    assert s.intersection(u) == intspan('8')

    assert s & t == s.intersection(t)
    assert t & u == t.intersection(u)
    assert s & u == s.intersection(u)


def test_difference():
    s = intspan('1-8')
    t = intspan('2-5')
    assert s.difference(t) == intspan('1,6-8')
    assert t.difference(s) == intspan()

    assert s - t == s.difference(t)
    assert t - s == t.difference(s)


def test_symmetric_difference():
    s = intspan('1-8')
    t = intspan('2-5')
    assert s.symmetric_difference(t) == intspan('1,6-8')
    assert t.symmetric_difference(s) == intspan('1,6-8')
    assert t.symmetric_difference(t) == intspan()

    assert s ^ t == s.symmetric_difference(t)
    assert t ^ s == t.symmetric_difference(s)
    assert t ^ t == t.symmetric_difference(t)


def test_augmented_assignments():
    s = intspan('50-60')
    s |= intspan('10-20')
    assert s == intspan('10-20,50-60')
    s &= intspan('0-55')
    assert s == intspan('10-20,50-55')
    s -= intspan('16-20')
    assert s == intspan('10-15,50-55')
    s ^= intspan('10,99')
    assert s == intspan('11-15,50-55,99')

    t = intspan('50-60')
    t.update('10-20')
    assert t == intspan('10-20,50-60')
    t.intersection_update('0-55')
    assert t == intspan('10-20,50-55')
    t.difference_update('16-20')
    assert t == intspan('10-15,50-55')
    t.symmetric_difference_update('10,99')
    assert t == intspan('11-15,50-55,99')


def test_pop():
    s = intspan('100-110')
    assert s.pop() == 100
    assert s.pop() == 101
    assert s.pop() == 102
    assert s.pop() == 103
    assert s.pop() == 104
    assert s.pop() == 105
    assert s == intspan('106-110')

    s = intspan('1-2')
    assert s.pop() == 1
    assert s.pop() == 2
    with pytest.raises(KeyError):
        s.pop()


def test_ranges():
    assert intspan().ranges() == []
    assert intspan('2').ranges() == [(2, 2)]
    assert intspan('1-3').ranges() == [(1, 3)]
    assert intspan('1-3,5-6').ranges() == [(1, 3), (5, 6)]


def test_from_range():
    assert intspan.from_range(1, 3) == intspan('1-3')
    assert intspan.from_range(2, 44) == intspan('2-44')


def test_from_ranges():
    assert intspan.from_ranges([(1, 3), (5, 6)]) == intspan('1-3,5-6')
    assert intspan.from_ranges([(1, 3)]) == intspan('1-3')
    assert intspan.from_ranges([(2, 2)]) == intspan('2')
    assert intspan.from_ranges([]) == intspan()


def test_universe():
    assert intspan().universe() == intspan()
    assert intspan('').universe() == intspan()
    assert intspan([]).universe() == intspan()

    assert intspan('1').universe() == intspan('1')

    assert intspan('1,3,5,7').universe() == intspan('1-7')

    s = intspan('1,3,5-9')
    assert s.universe() == intspan('1-9')
    assert s.universe(high=10) == intspan('1-10')
    assert s.universe(high=14) == intspan('1-14')
    assert s.universe(low=0) == intspan('0-9')
    assert s.universe(low=0, high=14) == intspan('0-14')
    assert s.universe(-2, 5) == intspan('-2-5')

    assert intspan('1-100').universe() == intspan('1-100')


def test_complement():
    s = intspan('1,3,5-9')
    assert s.complement() == intspan('2,4')
    assert s.complement(high=10) == intspan('2,4,10')
    assert s.complement(high=14) == intspan('2,4,10-14')
    assert s.complement(low=0) == intspan('0,2,4')
    assert s.complement(low=0, high=14) == intspan('0,2,4,10-14')

    assert s.complement(-2, 5) == intspan('-2,-1,0,2,4')

    items = intspan('1-3,5,7-9,10,21-24')
    assert items.complement() == intspan('4,6,11-20')
    assert items.complement(high=30) == intspan('4,6,11-20,25-30')

    with pytest.raises(ValueError):
        intspan().complement()
        # cannot get the complement of an empty set


def test_repr_and_str():
    s = intspan('10-20,50-55')
    s.add(9)
    s.discard('15-40')
    assert str(s) == '9-14,50-55'
    assert repr(s) == "intspan('" + str(s) + "')"


intspan.py

import sys
import copy
from itertools import groupby, count, chain
import re

__all__ = 'intspan spanlist intspanlist TheRest'.split()

_PY3 = sys.version_info[0] > 2
if _PY3:
    basestring = str


# Define regular expressions for spans and spans that may contain
# star (TheRest) markers
SPANRE = re.compile(r'^\s*(?P<start>-?\d+)\s*(-\s*(?P<stop>-?\d+))?\s*$')
SPANRESTAR = re.compile(
    r'^\s*((?P<star>\*)|(?P<start>-?\d+)\s*(-\s*(?P<stop>-?\d+))?)\s*$')


class ParseError(ValueError):
    pass


class Rester(object):

    """
    Singleton to represent "the rest of the values."
    """

    def __repr__(self):
        return 'TheRest'

    def __str__(self):
        return '*'


TheRest = Rester()


def _parse_range(datum):

    """
    Parser for intspan and intspan list.
    """

    def parse_chunk(chunk):
        """
        Parse each comma-separated chunk. Hyphens (-) can indicate ranges,
        or negative numbers. Returns a list of specified values. NB Designed
        to parse correct input correctly. Results of incorrect input are
        undefined.
        """
        m = SPANRESTAR.search(chunk)
        if m:
            if m.group('star'):
                return [TheRest]
            start = int(m.group('start'))
            if not m.group('stop'):
                return [start]
            stop = int(m.group('stop'))
            return list(range(start, stop + 1))
        else:
            raise ParseError("Can't parse chunk '{0}'".format(chunk))

    if isinstance(datum, basestring):
        result = []
        for part in datum.split(','):
            chunk = part.strip()
            if chunk:
                result.extend(parse_chunk(chunk))
        return result
    else:
        return datum if hasattr(datum, '__iter__') else [datum]


def spanlist(spec=None, chunkre=SPANRESTAR):
    """
    Given a string specification like the ones given to ``intspan``,
    return a list of the included items, in the same item given. Thus,
    ``spanlist("3,1-4")`` yields ``[3, 1, 2, 4]``. Experimental partial
    implementation of ability to have ordered intspans.
    """
    if spec is None or (isinstance(spec, basestring) and spec.strip() == ''):
        return []
    rawitems = _parse_range(spec)
    seen = set()
    items = []
    for i in rawitems:
        if i in seen:
            continue
        items.append(i)
        seen.add(i)
    return items


def _as_range(iterable):
    """
    Return a tuple representing the bounds of the range.
    """
    l = list(iterable)
    return (l[0], l[-1])


def _as_range_str(iterable):
    """
    Return a string representing the range as a string span.
    """
    l = list(iterable)
    if len(l) > 1:
        return '{0}-{1}'.format(l[0], l[-1])
    else:
        return '{0}'.format(l[0])


def _noRestDiff(a, b):
    """
    Special difference that, in case difference cannot be computed
    because of ``TypeError`` (indicating that a ``Rester`` object)
    has been found), returns a difference indicating the spanned items
    / group has ended.
    """
    try:
        return a - b
    except TypeError:
        return 2  # anything more than 1 signals "the next thing is in
                  # another span, not this current one"


class intspan(set):

    """
    A set of integers, expressed as an ordered sequence of spans.
    Because ``intspan('1-3,14,29,92-97')`` is better than
    ``[1, 2, 3, 14, 29, 92, 93, 94, 95, 96, 97]``.
    """

    def __init__(self, initial=None):
        """
        Construct a new ``intspan``.

        :param iterable|str initial: Optional initial list of items.
        :returns: the intspan
        :rtype: intspan
        """
        super(intspan, self).__init__()
        if initial:
            self.update(initial)

    def copy(self):
        """
        Return a new set with the same members.
        """
        return copy.copy(self)

    def update(self, items):
        """
        Add multiple items.

        :param iterable|str items: Items to add. May be an intspan-style string.
        """
        super(intspan, self).update(_parse_range(items))
        return self

    def intersection_update(self, items):
        super(intspan, self).intersection_update(_parse_range(items))
        return self

    def difference_update(self, items):
        super(intspan, self).difference_update(_parse_range(items))
        return self

    def symmetric_difference_update(self, items):
        super(intspan, self).symmetric_difference_update(
            _parse_range(items))
        return self

    def discard(self, items):
        """
        Discard items.

        :param iterable|str items: Items to remove. May be an intspan-style string.
        """
        for item in _parse_range(items):
            super(intspan, self).discard(item)

    def remove(self, items):
        for item in _parse_range(items):
            super(intspan, self).remove(item)

    def add(self, items):
        """
        Add items.

        :param iterable|str items: Items to add. May be an intspan-style string.
        """
        for item in _parse_range(items):
            super(intspan, self).add(item)

    def issubset(self, items):
        return super(intspan, self).issubset(_parse_range(items))

    def issuperset(self, items):
        return super(intspan, self).issuperset(_parse_range(items))

    def union(self, items):
        return intspan(super(intspan, self).union(_parse_range(items)))

    def intersection(self, items):
        return intspan(super(intspan, self).intersection(_parse_range(items)))

    def difference(self, items):
        return intspan(super(intspan, self).difference(_parse_range(items)))

    def symmetric_difference(self, items):
        return intspan(super(intspan, self).symmetric_difference(_parse_range(items)))

    __le__   = issubset
    __ge__   = issuperset
    __or__   = union
    __and__  = intersection
    __sub__  = difference
    __xor__  = symmetric_difference
    __ior__  = update
    __iand__ = intersection_update
    __isub__ = difference_update
    __ixor__ = symmetric_difference_update

    def __eq__(self, items):
        return super(intspan, self).__eq__(_parse_range(items))

    def __lt__(self, items):
        return super(intspan, self).__lt__(_parse_range(items))

    def __gt__(self, items):
        return super(intspan, self).__gt__(_parse_range(items))

    def __iter__(self):
        """
        Iterate in ascending order.
        """
        return iter(sorted(super(intspan, self).__iter__()))

    def pop(self):
        """
        Remove and return an arbitrary element; raises KeyError if empty.

        :returns: Arbitrary member of the set (which is removed)
        :rtype: int
        :raises KeyError: If the set is empty.

        """
        if self:
            min_item = min(self)
            self.discard(min_item)
            return min_item
        else:
            raise KeyError('pop from an empty set')

        # This method added only for PyPy, which otherwise would get the wrong
        # answer (unordered).


    def universe(self, low=None, high=None):
        """
        Return the "universe" or "covering set" of the given intspan--that
        is, all of the integers between its minimum and missing values.
        Optionally allows the bounds of the universe set to be manually specified.

        :param int low: Low bound of universe.
        :param int high: High bound of universe.
        :returns: the universe or covering set
        :rtype: intspan
        """
        if not self and low is None and high is None:
            return intspan()
        low = low if low is not None else min(self)
        high = high if high is not None else max(self)
        universe = self.__class__.from_range(low, high)
        return universe

    def complement(self, low=None, high=None):
        """
        Return the complement of the given intspan--that is, all of the
        'missing' elements between its minimum and missing values.
        Optionally allows the universe set to be manually specified.

        :param int low: Low bound of universe to complement against.
        :param int high: High bound of universe to complement against.
        :returns: the complement set
        :rtype: intspan
        :raises ValueError: if the set is empty (thus the compelement is infinite)
        """
        if not self:
            raise ValueError('cannot represent infinite set')
        return self.universe(low, high) - self

    @classmethod
    def from_range(cls, low, high):
        """
        Construct an intspan from the low value to the high value,
        inclusive. I.e., closed range, not the more typical Python
        half-open range.

        :param int low: Low bound of set to construct.
        :param int high: High bound of set to construct.
        :returns: New intspan low-high.
        :rtype: intspan
        """
        return cls(range(low, high + 1))

    @classmethod
    def from_ranges(cls, ranges):
        """
        Construct an intspan from a sequence of (low, high) value
        sequences (lists or tuples, say). Note that these values are
        inclusive, closed ranges, not the more typical Python
        half-open ranges.

        :param list ranges: List of closed/inclusive ranges, each a tuple.
        :returns: intspan combining the ranges
        :rtype: intspan
        """
        return cls(chain(*(range(r[0], r[1] + 1) for r in ranges)))

    def __repr__(self):
        """
        Return the representation.
        """
        clsname = self.__class__.__name__
        return '{0}({1!r})'.format(clsname, self.__str__())

    def __str__(self):
        """
        Return the stringification.
        """
        items = sorted(self)
        gk = lambda n, c=count(): n - next(c)
        return ','.join(_as_range_str(g) for _, g in groupby(items, key=gk))

    def ranges(self):
        """
        Return a list of the set's contiguous (inclusive) ranges.

        :returns: List of all contained ranges.
        :rtype: list
        """
        items = sorted(self)
        gk = lambda n, c=count(): n - next(c)
        return [_as_range(g) for _, g in groupby(items, key=gk)]

    # see Jeff Mercado's answer to http://codereview.stackexchange.com/questions/5196/grouping-consecutive-numbers-into-ranges-in-python-3-2
    # see also: http://stackoverflow.com/questions/2927213/python-finding-n-consecutive-numbers-in-a-list


# It might be interesting to have a metaclass factory that could create
# spansets of things other than integers. For example, enumerateds defined
# by giving a universe of possible options. Or characters. The Ranger
# package seems to do some of this http://pythonhosted.org/ranger/


class intspanlist(list):

    """
    An ordered version of ``intspan``. Is to ``list`` what ``intspan``
    is to ``set``, except that it is somewhat set-like, in that items
    are not intended to be repeated. Works fine as an immutable
    data structure. Still some issues if one mutates an instance. Not
    terrible problems, but the set-like nature where there is only
    one entry for each included integer may be broken.
    """

    def __init__(self, initial=None, universe=None):
        """
        Construct a new ``intspanlist``

        :param iterable|str initial: Optional initial list of items.
        :returns: the intspanlist
        :rtype: intspanlist
        """
        super(intspanlist, self).__init__()
        if initial:
            self.extend(initial)
        if universe is not None:
            try:
                restIndex = self.index(TheRest)
                remaining = sorted(intspan(universe) - set(self))
                self[restIndex + 1:restIndex + 1] = remaining  # splice
                self.pop(restIndex)
            except ValueError:
                pass

    def therest_update(self, universe, inplace=True):
        """
        If the receiving ``intspanlist`` contains a ``TheRest`` marker,
        replace it with the contents of the universe. Generally done
        *in situ*, but if value of ``inplace`` kwarg false, returns
        an edited copy.
        """
        toedit = self if inplace else self.copy()
        try:
            restIndex = toedit.index(TheRest)
            remaining = sorted(intspan(universe) - set(toedit))
            toedit[restIndex + 1:restIndex + 1] = remaining  # splice
            toedit.pop(restIndex)
        except ValueError:
            pass
        return toedit

    def copy(self):
        """
        Return a copy of the intspanlist.
        """
        return copy.copy(self)

    def append(self, item):
        """
        Add to the end of the intspanlist

        :param int item: Item to add
        """
        self.extend(spanlist(item))

    def extend(self, items):
        """
        Add a collection to the end of the intspanlist

        :param iterable items: integers to add
        """
        seen = set(self)
        for newitem in spanlist(items):
            if newitem in seen:
                continue
            super(intspanlist, self).append(newitem)

    def __eq__(self, items):
        return super(intspanlist, self).__eq__(spanlist(items))

    def __lt__(self, items):
        return super(intspanlist, self).__lt__(spanlist(items))

    def __gt__(self, items):
        return super(intspanlist, self).__gt__(spanlist(items))

    def complement(self, low=None, high=None):
        """
        Return the complement of the given intspanlist--that is, all of the
        'missing' elements between its minimum and missing values.
        Optionally allows the universe set to be manually specified.

        :param int low: Low bound of universe to complement against.
        :param int high: High bound of universe to complement against.
        :returns: the complement set
        :rtype: intspanlist
        :raises ValueError: if the set is empty (thus the compelement is infinite)
        """
        cls = self.__class__
        if not self:
            raise ValueError('cannot represent infinite set')
        low = low if low is not None else min(self)
        high = high if high is not None else max(self)
        universe = cls.from_range(low, high)
        result = []
        contained = set(self)
        for x in universe:
            if x not in contained:
                result.append(x)
        return cls(result)

    @classmethod
    def from_range(cls, low, high):
        """
        Construct an intspanlist from the low value to the high value,
        inclusive. I.e., closed range, not the more typical Python
        half-open range.

        :param int low: Low bound.
        :param int high: High bound.
        :returns: New intspanlist low-high.
        :rtype: intspanlist
        """
        return cls(range(low, high + 1))

    @classmethod
    def from_ranges(cls, ranges):
        """
        Construct an intspanlist from a sequence of (low, high) value
        sequences (lists or tuples, say). Note that these values are
        inclusive, closed ranges, not the more typical Python
        half-open ranges.


        :param list ranges: List of closed/inclusive ranges, each a tuple.
        :returns: intspanlist combining the ranges
        :rtype: intspanlist
        """
        return cls(chain(*(range(r[0], r[1] + 1) for r in ranges)))

    def __repr__(self):
        """
        Return the representation.
        """
        clsname = self.__class__.__name__
        return '{0}({1!r})'.format(clsname, self.__str__())

    def __str__(self):
        """
        Return the stringification.
        """
        gk = lambda n, c=count(): _noRestDiff(n, next(c))
        return ','.join(_as_range_str(g) for _, g in groupby(self, key=gk))

    def ranges(self):
        """
        Return a list of the set's contiguous (inclusive) ranges.
        """
        gk = lambda n, c=count(): _noRestDiff(n, next(c))
        return [_as_range(g) for _, g in groupby(self, key=gk)]
